x/bank/types/balance.go

```diff
@@ -34,6 +34,11 @@ func (b Balance) Validate() error {
 	if err != nil {
 		return err
 	}
+
+	var prevDenom string
```

Comment:
There is no need to do this. `Coins` already has a `Validate` method that will return `false` if they're not sorted.

Reply:
```go
if err := b.Coins.Validate(); err != nil {
  return err
}
```

Reply:
The comment on the function specifies the need for custom validation instead of using `Coins.Validate`
https://github.com/cosmos/cosmos-sdk/blob/2cea5e5750d220d476ddc3184d19cdf0a7de52c5/x/bank/types/balance.go#L44-L45

Reply:
That's fine I think, it then becomes

```go
if err := b.Coins.Validate(); err != nil && !b.Coins.IsEmpty() {
  return err
}
```

Reply:
Under the current implementation, a case like this validates to `true`. 
```go
bank.Balance{
	Address: "cosmos1yq8lgssgxlx9smjhes6ryjasmqmd3ts2559g0t",
	Coins: sdk.Coins{
		sdk.NewInt64Coin("atom", 0),
		sdk.NewInt64Coin("zatom", 2),
	},
},
```
Using `Coins.Validate` does not allow for zero amount for coin and this case would validate to `false`.

Reply:
Yes, that's correct. Zero coins are not allowed.

Reply:
Ah okay. I understood the comment to mean that `sdk.NewInt64Coin("atom", 0)` should be allowed and thats why custom validation was being done.

Reply:
Yeah, sorry for the confusion! So yes, zero-value coins are not allowed anywhere in the state-machine, they're not valid 👍